src/MongoDB.Driver.Core/Core/Authentication/ScramShaAuthenticator.cs

```diff
@@ -214,19 +223,34 @@ public ISaslStep Transition(SaslConversation conversation, byte[] bytesReceivedF
                 var nonce = "r=" + r;
                 var clientFinalMessageWithoutProof = channelBinding + "," + nonce;
 
-                var saltedPassword = _hi(
-                    _credential,
-                    Convert.FromBase64String(s),
-                    int.Parse(i));
+                byte[] clientKey;
+                byte[] serverKey;
+                //placeholder until I know what to use for hashedPassword
+                byte[] hashedPassword = null;
+
+                var cacheKey = new ScramCacheKey(hashedPassword, Convert.FromBase64String(s), int.Parse(i));
+
+                if (_cache.Get(cacheKey) != null)
+                {
+                    clientKey = _cache.Get(cacheKey).ClientKey;
+                    serverKey = _cache.Get(cacheKey).ServerKey;
+                }
+                else
+                {
+                    var saltedPassword = _hi(
+                        _credential,
+                        Convert.FromBase64String(s),
+                        int.Parse(i));
+                    clientKey = _hmac(encoding, saltedPassword, "Client Key");
+                    serverKey = _hmac(encoding, saltedPassword, "Server Key");
+                    _cache.Set(cacheKey, new ScramCacheEntry(clientKey, serverKey));
+                }
```

Comment:
Replace lines 227-247 with:

    var cacheKey = new ScramCacheKey(_credential.SaslPreppedPassword, salt, iterations);
    if (_cache.TryGet(cacheKey, out var cacheEntry))
    {
        clientKey = cacheEntry.ClientKey;
        serverKey = cacheEntry.ServerKey;
    }
    else
    {
        var saltedPassword = _hi( _credential, salt, iterations);
        clientKey = _hmac(encoding, saltedPassword, "Client Key");
        serverKey = _hmac(encoding, saltedPassword, "Server Key");
        _cache.Set(cacheKey, new ScramCacheEntry(clientKey, serverKey));
    }

This depends on changing the type of the password parameter to ScramCacheKey to SecureString.

Reply:
Done